wxWebRequest build fixes

MaartenBent opened the pull request:
See https://trac.wxwidgets.org/ticket/19105

MaartenBent committed:
Check if wxUSE_WEBREQUEST is defined

MaartenBent committed:
Fix checking __MINGW64_VERSION_MAJOR for wxUSE_WEBREQUEST_WINHTTP

It is not yet defined for setup.h, so check it in chkconf.h.
Fixes #19105

MaartenBent commented:
So besides the header, MinGW also does not have the `winhttp` library. Any ideas how to work around this?
Or is the only solution to load the WinHttp functions dynamically, like was once done for uxtheme (and removed in ddceaab001a6c9459af84f3937e99fa9b8dc755a).

TcT2k commented:
> So besides the header, MinGW also does not have the `winhttp` library. Any ideas how to work around this?
> Or is the only solution to load the WinHttp functions dynamically, like was once done for uxtheme (and removed in [ddceaab](https://github.com/wxWidgets/wxWidgets/commit/ddceaab001a6c9459af84f3937e99fa9b8dc755a)).

Does it at least contain the `winhttp` headers?
I think a difference between this and uxtheme is that winhttp is available on every supported windows (which uxtheme was not at the pre XP times). So this should be handled by a helper macros/include/cpp as this is just to support an outdated Windows SDK (and only to be used in that case).

vadz commented:
Hmm, I see `libwinhttp.a` as far back as MinGW-w64 4.8 and I also see it in TDM 5.1 and MSYS2. Which version exactly is missing it?

MaartenBent commented:
The one used by appveyor, I think https://sourceforge.net/projects/mingw/: https://ci.appveyor.com/project/wxWidgets/wxwidgets/builds/38296664/job/c3uu2poxva4xkxhe#L1271

This build already has webrequest disabled with the `__MINGW64_VERSION_MAJOR` check. But `-lwinhttp` is now in the library list, so it fails when it is not found, even if it is not used.

Unrelated: I forgot checking for `_MSC_VER` in `chkconf.h`.

vadz commented:
Is this actually a MinGW-w64 at all? This looks like plain MinGW 5.3 to me and it doesn't have neither the header nor the library, as expected. But we shouldn't use this backend with non-w64 MinGW...

MaartenBent commented:
It is indeed plain MinGW 5.3. And `wxWebRequest` is disabled. But the generated makefiles have `-lwinhttp` in it. That causes the build to fail.
I do not think there is a way in bakefile to link with winhttp only for specific MinGW compilers. So I'll load the functions dynamically.

MaartenBent committed:
Load winhttp functions dynamically

winhttp cannot be added to the makefiles, because some MinGW distributions
(MinGW 5.3) do not have this library.

vadz commented:
I'm not against it, but it seems a pity to do all the work for dynamic linking just to support plain MinGW that nobody uses any longer :-( Perhaps we could add `USE_WINHTTP` make variable (i.e. to `build/msw/config.gcc`) that would need to be set to 1 to use it? This would inconvenience people using MinGW-w64 (without configure, at least, which is how I always personally use it), but would be much simpler.

MaartenBent commented:
I think quite a few people use the gcc makefiles, also the reporter of the ticket. So manually having to enable it seems not usable to me. And I've already written the dynamic loader.

MaartenBent committed:
Fix extra semi warning in wxWebSessionCURL implementation

MaartenBent commented:
The ASAN build error seems unrelated. Any other comments regarding this solution? Otherwise it could be merged.

vadz commented on src/osx/webrequest_urlsession.mm:
```diff
@@ -12,7 +12,7 @@
 #include "wx/webrequest.h"
 
-#if wxUSE_WEBREQUEST_URLSESSION
+#if wxUSE_WEBREQUEST && wxUSE_WEBREQUEST_URLSESSION
```
I think `wx/chkconf.h` should check that `wxUSE_WEBREQUEST==1` if `wxUSE_WEBREQUEST_URLSESSION==1`.

MaartenBent replied:
I suppose it is possible to use curl on macOS and disable URLSession, this check would prevent that.

vadz replied:
Ah, yes, you're right, of course, sorry.

vadz reviewed (commented):
Thanks for the ping, I've finally looked at it and while globally this is exactly the right thing to do, of course, I have a lingering concern about loading the DLL on startup. Maybe I'm overthinking things here, but OTOH loading it from ` wxWebSession::InitFactoryMap()` should do no harm and seems simple to do.

Thanks again!

P.S. Oops, my comment somehow didn't get published :-( Let me retype it and submit again.

P.P.S. Resubmitted again, it was the comment for the module `OnInit()` which got somehow lost the first time. Sorry for the confusion.

vadz commented on src/msw/webrequest_winhttp.cpp:
```diff
@@ -12,23 +12,121 @@
+public:
+    virtual bool OnInit() wxOVERRIDE
+    {
+        return wxWinHTTP::LoadLibraries();
```
I don't know if this is going to really be a problem in practice, but failing to load `winhttp.dll`, for whatever reason, will now prevent any wx application (linking with wxnet) from starting up, even if it doesn't use `wxWebRequest` at all. Again, this might well never happen, but it still seems a bit worrisome.

It would be better to do this on demand only, but I'm not sure if we know which function will always be called first. Can we rely on no WinHTTP functions being called before `wxWebRequestWinHTTP` ctor?

Also, there is still a risk of crashing if we fail and the function pointer is null.

The only robust solution I see is to make all the function wrappers, i.e. check if it could be loaded and return an error otherwise, but this would require even more boilerplate code.

So, finally, the best solution I see is to do this from `wxWebSession::InitFactoryMap()`, i.e. add a check for successful initialization there before creating `wxWebSessionFactoryWinHTTP`. This should keep things reasonably simple, while ensuring that we don't do anything worse than not allowing to use WinHTTP if we fail to load the DLL or resolve any functions in it.

What do you think?

MaartenBent replied:
I agree, I'll move it to the initialization of `wxWebSessionFactoryWinHTTP` and check for success.

MaartenBent committed:
Initialize wxWinHTTP when wxWebSessionFactoryWinHTTP is created

vadz commented:
Perfect, thanks a lot!
